upup/pkg/fi/vfs_castore.go

```diff
@@ -203,88 +207,69 @@ func serializeKeysetBundle(o *kops.Keyset) ([]byte, error) {
 
 // removePrivateKeyMaterial returns a copy of the Keyset with the private key data removed
 func removePrivateKeyMaterial(o *kops.Keyset) *kops.Keyset {
-	copy := o.DeepCopy()
+	c := o.DeepCopy()
 
-	for i := range copy.Spec.Keys {
-		copy.Spec.Keys[i].PrivateMaterial = nil
+	for i := range c.Spec.Keys {
+		c.Spec.Keys[i].PrivateMaterial = nil
 	}
 
-	return copy
-}
-
-func SerializeKeyset(o *kops.Keyset) ([]byte, error) {
-	var objectData bytes.Buffer
-	{
-		codecs := kopscodecs.Codecs
-		yaml, ok := runtime.SerializerInfoForMediaType(codecs.SupportedMediaTypes(), "application/yaml")
-		if !ok {
-			klog.Fatalf("no YAML serializer registered")
-		}
-		encoder := codecs.EncoderForVersion(yaml.Serializer, v1alpha2.SchemeGroupVersion)
-
-		if err := encoder.Encode(o, &objectData); err != nil {
-			return nil, fmt.Errorf("error serializing keyset: %v", err)
-		}
-	}
-
-	return objectData.Bytes(), nil
-}
-
-func (c *VFSCAStore) loadCertificates(p vfs.Path) (*keyset, error) {
-	bundlePath := p.Join("keyset.yaml")
-	bundle, err := c.loadKeysetBundle(bundlePath)
-	return bundle, err
+	return c
 }
 
-func (c *VFSCAStore) loadOneCertificate(p vfs.Path) (*pki.Certificate, error) {
-	data, err := p.ReadFile()
-	if err != nil {
-		if os.IsNotExist(err) {
-			return nil, nil
-		}
-		return nil, err
-	}
-	cert, err := pki.ParsePEMCertificate(data)
-	if err != nil {
-		return nil, err
-	}
-	if cert == nil {
-		return nil, nil
-	}
-	return cert, nil
+func (c *VFSCAStore) FindKeypair(name string) (*pki.Certificate, *pki.PrivateKey, error) {
+	return FindKeypair(c, name)
 }
 
-func (c *VFSCAStore) FindKeypair(id string) (*pki.Certificate, *pki.PrivateKey, bool, error) {
-	cert, legacyFormat, err := c.findCert(id)
+func (c *VFSCAStore) FindKeyset(id string) (*Keyset, error) {
+	certs, err := c.loadKeyset(c.buildCertificatePoolPath(id))
```

Comment:
So technically we should be able to load just the certificates, even if we don't have permission to the keys.  I had a look, and the reason we get away with this (I think) is that AWS/S3 is the only cloud where we lock down access so most nodes can't access the keys, and on AWS we default to using the kops-controller.

Not sure if we want to address this or just move everyone to kops-controller and work towards not writing the certificate-only forms.  It wouldn't be too hard though I don't think - probably the easiest way is just to create FindKeysetCertificates and check each invocation of FindKeyset to see if it needs the certificates.

Also, I like the merge logic and we may want to keep it, but the way we avoid problems here is that we should always write the private keys first.  From the private keys, we can get the public keys / certs (but not vice versa) so that's why we do it in that order.  So it _should_ suffice to just load the private keys, assuming we have permission.

Reply:
Things that just want the certificates should get them through userdata or equivalent, not VFS. That way when the certificates change the thing gets updated by rolling update.

In any case, everything that currently calls `FindKeyset` (or `FindPrimaryKeypair`) has or needs access to the private key.

Reply:
The merge logic mimics the behavior of the previous `FindKeypair()`. A future PR could remove the certificate-only bundle entirely once it ensures that all certificate-only consumers have been refactored to use userdata or equivalent.